File: Папа и Мама/PapaMama/PapaMama.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading;

namespace Папа_мама
{
    class PapaMama
    {
        public static int num_of_repeats;
        private int _delay;
        private int _repeats;
        private string _message;
        public static object lo = new object();
        public PapaMama(string message, int repeats, int delay)
        {
            this._delay = delay;
            this._repeats = repeats;
            this._message = message;
        }

        public void Say()
        {
            int i = 0;
            while (true)
            {
                lock (lo)
                {
                    if (i < _repeats & num_of_repeats > 0)
                    {
                        Console.WriteLine(_message);
                        num_of_repeats--;
                    }
                }
                if (i >= _repeats | num_of_repeats <= 0)
                    break;
                i++;
                Thread.Sleep(_delay);
            }
        }
    }
}

File: Папа и Мама/PapaMama C++/PapaMama.cpp
//
//  PapaMama.cpp
//  папа и мама
//
//  Created by Владимир on 15.04.21.
//

#include "PapaMama.hpp"
#include<thread>
#include<iostream>
#include<mutex>

int PapaMama::num_of_repeats;
mutex mtx;

PapaMama::PapaMama(string message, int repeats, int delay)
{
    _message = message;
    _repeats = repeats;
    _delay = delay;
}

void PapaMama::operator()()
{
    int i = 0;
    mtx.lock();
    while (i < _repeats && num_of_repeats > 0)
    {
        cout << _message << endl;
        num_of_repeats--;
        mtx.unlock();
        i++;
        this_thread::sleep_for(chrono::milliseconds{_delay});
        mtx.lock();
    }
    mtx.unlock();
}
